src/network/cni/namespace.rs

```diff
@@ -0,0 +1,119 @@
+//! Network namespace helpers and structures.
+
+use anyhow::{Context, Result};
+use futures::executor;
+use getset::Getters;
+use log::trace;
+use nix::sched::{setns, CloneFlags};
+use std::{
+    fs,
+    future::Future,
+    os::unix::io::{AsRawFd, RawFd},
+    path::{Path, PathBuf},
+};
+use tokio::{fs::File, task};
+
+#[derive(Debug, Getters)]
+/// A basic network namespace abstraction.
+pub struct Namespace {
+    #[get]
+    /// The current namespace as File.
+    current: File,
+
+    #[get]
+    /// The target namespace as File.
+    target: File,
+}
+
+impl Namespace {
+    /// Create a new namespace.
+    pub async fn new<P>(path: P) -> Result<Self>
+    where
+        P: AsRef<Path>,
+    {
+        Ok(Self {
+            current: File::open(Self::current_thread_namespace_path())
+                .await
+                .context("open current thread namespace file")?,
+            target: File::open(&path)
+                .await
+                .context("open target namespace file")?,
+        })
+    }
+
+    /// Run a future inside this network namespace
+    pub async fn run<F>(&self, fun: F) -> Result<()>
+    where
+        F: Future<Output = Result<()>> + Send + 'static,
+    {
+        trace!(
+            "Using file as target network namespace: {:?}",
+            self.target()
+        );
+        let current_fd = self.current().as_raw_fd();
+        let target_fd = self.target().as_raw_fd();
+
+        task::spawn_blocking(move || {
+            // Switch to the target namespace
+            trace!("Switching to target namespace");
+            Self::switch_namespace(target_fd)?;
+
+            // Run the future
+            let result = executor::block_on(fun).context("run namespace future");
+
+            // Ensure that we will switch back to the original network namespace
+            trace!("Switching back to host network namespace");
+            Self::switch_namespace(current_fd)?;
+
+            result
+        })
```

Comment:
Okay, this should now not poison the network namespace of the thread any more.

Reply:
Yeah, looks reasonable now. :smile_cat:

Reply:
Thanks so much for the help! :pray: